feat(cognito): allow to set read and write attributes in Cognito UserPoolClient (#7607)
aws/aws-cdk: packages/@aws-cdk/aws-cognito/README.md

});
```

> For each client you should specify wich user attributes it will be able to read or write.

Reviewer: ```suggestion
Clients can be configured so that only specific user attributes can be read or modified by it.
```

> The default behavior is to allow read and write of all attribute, but it is often not the best option.

Reviewer: ```suggestion
The default behavior is to allow read and write all available attributes, but it is often not the best option.
```

> For example, in a webapp there might be a `customerPlan` custom attribute to store the user's plan (`free`, `silver` 
> or `gold`) but only some clients (i.e. those used by the invoice system) should be allowed to change its value.

Reviewer: ```suggestion
For example, consider a webapp that contains a custom attribute `customerPlan` to store the user's pricing plan
where only some clients, such as the administrators, should be allowed to change this value. Read more about [Attribute
Permissions and Scopes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes).
```

```ts
const pool = new UserPool(this, 'Pool');
pool.addClient('webapp-client', {
  readAttributes: undefined, // by default all attributes are readable/writable
  writeAttributes: {
    standard: [StandardAttribute.NICKNAME],
    custom: ['bio'],
  }, // only the nickname and the bio are writable
});

> ```

Reviewer: Can you modify this example to align with the pricing plan example you have above?

The default behavior is to allow read and write all available attributes, but it is often not the best option.
For example, consider a webapp that contains a custom attribute `customerPlan` to store the user's pricing plan
where only some clients, such as the administrators, should be allowed to change this value. Read more about [Attribute
Permissions and Scopes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes).

```ts
const pool = new UserPool(this, 'Pool');
pool.addClient('webapp-client', {
  readAttributes: {
    standard: [StandardAttribute.NICKNAME, StandardAttribute.EMAIL, StandardAttribute.NAME],
    custom: ['customerPlan'],
  },
  writeAttributes: {
    standard: [StandardAttribute.NICKNAME],
  }, // only the nickname will be writable
});
```

The default behavior of allowing read/write of all attributes is enforced at Client creation, and only those
attributes available at that time will be included.
If any attribute is added to the User Pool after the creation of the Client, that attribute will not be
available unless explicitly specified.

> 

Author: I've added those lines but I'm not sure it's easy to understand what i mean. @nija-at could you please give me a feedback?

Reviewer: This applies only for custom attributes, no? Because I thought standard attributes are always available.

});
```

Clients can (and should) be allowed to read and write relevant user attributes only. Usually every client can be allowed to read the `given_name`
attribute but not every client should be allowed to set the `email_verified` attribute.
The same criteria applies for both standard and custom attributes, more info is available at
[Attribute Permissions and Scopes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes).
The default behaviour is to allow read and write permissions on all attributes.

```ts
const pool = new cognito.UserPool(this, 'Pool');
pool.addClient('app-client', {
  // ...

>   readAttributes: AttributeSet.allStandard(['custom:my_attribute_1', 'custom:my_attribute_2']),

Reviewer: Given that we know that it is custom, we could avoid the 'custom' prefix?

Author: Sure… much cleaner that way.

});
```

Clients can (and should) be allowed to read and write relevant user attributes only. Usually every client can be allowed to read the `given_name`
attribute but not every client should be allowed to set the `email_verified` attribute.
The same criteria applies for both standard and custom attributes, more info is available at
[Attribute Permissions and Scopes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes).

> The default behaviour is to allow read and write permissions on all attributes.
> 
> ```ts
> const pool = new cognito.UserPool(this, 'Pool');
> // create a set of attributes that the client will be allowed to set
> const clientWriteAttributes = ClientAttributes.empty()
>   .withStandardAttributes({name: true, email: true})
>   .withCustomAttributes(['favouritePizza']);
> // read attributes are usually more than the writable ones
> const clientReadAttributes = clientWriteAttributes
>   .withStandardAttributes({emailVerified: true})
>   .withCustomAttributes(['pointsEarned']);
> pool.addClient('app-client', {
>   // ...
>   readAttributes: clientReadAttributes,
>   writeAttributes: clientWriteAttributes,
> });
> ```

Reviewer: Slight tweaks to the documentation.

````suggestion
The default behaviour is to allow read and write permissions on all attributes. The following code shows how this can be configured for a client.

```ts
const pool = new cognito.UserPool(this, 'Pool');

const clientWriteAttributes = ClientAttributes.empty()
  .withStandardAttributes({name: true, email: true})
  .withCustomAttributes(['favouritePizza']);

const clientReadAttributes = clientWriteAttributes
  .withStandardAttributes({emailVerified: true})
  .withCustomAttributes(['pointsEarned']);
pool.addClient('app-client', {
  // ...
  readAttributes: clientReadAttributes,
  writeAttributes: clientWriteAttributes,
});
```
````
